EKF/ekf_helper.cpp

```diff
@@ -1819,3 +1816,137 @@ void Ekf::stopFlowFusion()
 	memset(_flow_innov_var,0.0f,sizeof(_flow_innov_var));
 	memset(&_optflow_test_ratio,0.0f,sizeof(_optflow_test_ratio));
 }
+
+// Reset main nav filter yaw to value from EKF-GSF and reset velocity and position to GPS
+bool Ekf::resetYawToEKFGSF()
+{
+	// don't allow reet using the EKF-GSF estimate until the filter has started fusing velocity
+	// data and the yaw estimate has converged
+	float new_yaw, new_yaw_variance;
+	if (yawEstimator.getYawData(&new_yaw, &new_yaw_variance) && new_yaw_variance < sq(_params.EKFGSF_yaw_err_max)) {
+		// save a copy of the quaternion state for later use in calculating the amount of reset change
+		Quatf quat_before_reset = _state.quat_nominal;
+
+		// update transformation matrix from body to world frame using the current estimate
+		_R_to_earth = Dcmf(_state.quat_nominal);
+
+		// update the rotation matrix using the new yaw value
+		// determine if a 321 or 312 Euler sequence is best
+		if (fabsf(_R_to_earth(2, 0)) < fabsf(_R_to_earth(2, 1))) {
+			// use a 321 sequence
+			Eulerf euler321(_state.quat_nominal);
+			euler321(2) = new_yaw;
+			_R_to_earth = euler321;
+
+		} else {
+			// Calculate the 312 Tait-Bryan rotation sequence that rotates from earth to body frame
+			// We use a 312 sequence as an alternate when there is more pitch tilt than roll tilt
+			// to avoid gimbal lock
+			Vector3f rot312;
+			rot312(0) = new_yaw;
+			rot312(1) = asinf(_R_to_earth(2, 1));
+			rot312(2) = atan2f(-_R_to_earth(2, 0), _R_to_earth(2, 2));
+
+			float c2 = cosf(rot312(2));
+			float s2 = sinf(rot312(2));
+			float s1 = sinf(rot312(1));
+			float c1 = cosf(rot312(1));
+			float s0 = sinf(rot312(0));
+			float c0 = cosf(rot312(0));
+
+			_R_to_earth(0, 0) = c0 * c2 - s0 * s1 * s2;
+			_R_to_earth(1, 1) = c0 * c1;
+			_R_to_earth(2, 2) = c2 * c1;
+			_R_to_earth(0, 1) = -c1 * s0;
+			_R_to_earth(0, 2) = s2 * c0 + c2 * s1 * s0;
+			_R_to_earth(1, 0) = c2 * s0 + s2 * s1 * c0;
+			_R_to_earth(1, 2) = s0 * s2 - s1 * c0 * c2;
+			_R_to_earth(2, 0) = -s2 * c1;
+			_R_to_earth(2, 1) = s1;
+
+		}
+
+		// calculate the amount that the quaternion has changed by
+		Quatf quat_after_reset = _R_to_earth;
+		Quatf q_error =  quat_after_reset * quat_before_reset.inversed();
+		q_error.normalize();
+
+		// update quaternion states
+		_state.quat_nominal = quat_after_reset;
+		uncorrelateQuatFromOtherStates();
+
+		// record the state change
+		_state_reset_status.quat_change = q_error;
+
+		// update transformation matrix from body to world frame using the current estimate
+		_R_to_earth = Dcmf(_state.quat_nominal);
```

Comment:
This look the same as:
- https://github.com/PX4/ecl/blob/3fa5f501ae7edc4a793f0973e156706afcc8b419/EKF/ekf_helper.cpp#L688-L697
- https://github.com/PX4/ecl/blob/3fa5f501ae7edc4a793f0973e156706afcc8b419/EKF/gps_yaw_fusion.cpp#L366-L388
But not with the same exact order; is that by design? Can't we unify most of the "yaw reset" logic? Why are we computing 
```c++
float scalar; 
  
 if (q_error(0) >= 0.0f) { 
 	scalar = -2.0f; 
  
 } else { 
 	scalar = 2.0f; 
 } 
  
 delta_ang_error(0) = scalar * q_error(1); 
 delta_ang_error(1) = scalar * q_error(2); 
 delta_ang_error(2) = scalar * q_error(3); 
```
In `gps_yaw_fusion` but not in the other ones?

Reply:
Unification is something I would like to do, but would rather tackle it in a separate PR. The code we have for the current reset to EKF-GSF yaw is working reliably based on testing to date, so I would be reluctant to change it until we fully understand the consequences of doing so.

Reply:
The idea behind the delta_ang_error check was to reset the states and covariances only if the yaw change was significant and was intended to the used for the types of resets we do every flight, such as the in-air reset after takeoff when the vehicle has cleared groudn based magnetic interference.

In the case of an emergency yaw reset we work on the assumption that a hard reset is necessary so no need to do that check.